### ISSUE TITLE
Collaborate (read: steal ideas) with @bencates on system76/policy

### ISSUE THREAD

**Issue by schrockwell** [schrockwell/authy#5]
After discussion with @bencates at ElixirConf...

Primary change: additional controller helpers for `authorize`:

*  `authorize/3`: new, returns `{:ok, conn}` or `{:error, reason}` - good for handling `with` statements (a bit of future-proofing for Phoenix 1.3, and nice and Elixiry)
* `authorize!/3`: new, returns conn on success, raises on failure
* ~~`authorize/4`: tweak: it's still a macro, but let's explicitly pass in `conn` instead of hiding it~~

Also:
* Drop handler config, rely on `MyApp.ErrorView` to display unauthorized
* Don't care about 404 case – let `Repo.get!` raise and explain more complicated case in readme
* Add `:authy_authorized` marker to via `put_private`
* Add auth plug that utilizes `register_before_send` - call it `verify_authorized`
* Probably expose `mark_authorized` to explicitly skip authorization

**Comment by schrockwell** [comment]
@bencates I'd like to use the [git-flow](http://nvie.com/posts/a-successful-git-branching-model/) branching strategy for making changes. Basically we can do all active development on a `develop` branch, and merge changes over to `master` only when the package is finally published on Hex, tagging it with the version number. That way `master` will align with the most current release.

**Comment by schrockwell** [comment]
I think we should remove the `authorize` macro entirely, in part because the proposed `authorize/4` function will conflict with it, and in part because do-blocks are very Ruby-esque, and we should embrace `{:ok, result}` and the pipeline.

@bencates also has convinced me that raising exceptions is OK, since an authorization failure is truly an exceptional case. Also, Plug and Phoenix already have existing mechanisms for raising and catching exceptions, so we should leverage that and drop the custom callback config entirely.

This looks and feels a lot nicer, with the pipelining and the explicit passing of `conn` to the helpers.

```elixir
defmodule MyApp.PostController do
  use MyApp.Web, :controller
  alias MyApp.Post

  # Authy.Controller has been imported in web.ex

  def index(conn, _params) do
    conn
    |> authorize!(Post)                        # <-- authorize the :index action for Post
    |> render("index.html", posts: scope(conn, Post)) # <-- posts scoped per user
  end

  def show(conn, %{"id" => id}) do
    post = scope(conn, Post) |> Repo.get(id)   # <-- scope can be used for lookup

    conn
    |> authorize!(post)                        # <-- authorize the :show action for this particular post
    |> render("show.html", post: post)
  end

  def update(conn, %{"id" => id, "post" => post_params}) do
    post = scope(conn, Post) |> Repo.get(id)   # <-- scope can be used for lookup
    conn = authorize!(conn, post)              # <-- authorize the :update action for this particular post

    changeset = Post.changeset(post, post_params)
    case Repo.update(changeset) do
      # ...
    end
  end
end
```

**Comment by bencates** [comment]
So does `scope/2` return an `Ecto.Query` object or a list of schemas?  Because I don't think your index will work; needs to be `scope(conn, Post) |> Repo.all`.

**Comment by schrockwell** [comment]
Ah you're right. So it's not quite as clean as in that example. I was mainly trying to demonstrate the bang method usage to get a feel for it. 

You okay with dropping the old authorize macro in favor of the plain function that returns a tuple? Seems like the most idiomatic way to go about it.

**Comment by bencates** [comment]
Sure, drop away. Less API is always better.

**Comment by schrockwell** [comment]
Great. I've already made the change to `authorize/3` on `develop`, so I will wait on your PR for the bang methods before going further. We can do the rename and fix up the README as the final step.

**Comment by schrockwell** [comment]
This looks great. We've already hit pretty much everything on the list.

I'm undecided on how to handle the nil case. If we want to go the route of pure authorization, we should probably remove the `:nils` option entirely and either (a) raise an `ArgumentError` (will be a 500 error I think), or (b) raise our `Authy.NotAuthorizedError`. 

I'm leaning towards (a).

**Comment by bencates** [comment]
I've got some ideas there. I'll try to get another PR put together this
evening so we have something concrete to talk over.

**Comment by schrockwell** [comment]
👍 Make sure to pull down my latest changes.

**Comment by schrockwell** [comment]
Dumb question, but how do we wire up Phoenix to correctly catch the exception and render the `ErrorView` template? Check out the [Custom Errors](http://www.phoenixframework.org/docs/custom-errors) example at the bottom.

**Comment by bencates** [comment]
http://stackoverflow.com/questions/30536979/setting-up-custom-response-for-exception-in-phoenix-application

tl;dr run it in production mode

**Comment by schrockwell** [comment]
@bencates For now I've just "punted" the nil case, and removed the `nils` option entirely.

* If `nil` is passed as the data, authorization fails immediately, without deferring to a policy, since it can't be determined.
* if `nil` is passed with the `policy` option specified, then that policy module is used, passing `nil` as the data.

**Comment by schrockwell** [comment]
It looks like we've hit everything on this list, and I'm pretty happy with the result.

Please let me know if there's anything we missed. Otherwise I'll close this and move forward with the release.

**Comment by bencates** [comment]
I'm almost inclined to push `nil` handling down to `Bodyguard.authorized?/4`, and I'm comfortable with it just returning `false`.

That said, I'm okay with closing this and releasing.  I think we've got the controller API in a comfortable place.  I've got a PR with some code cleanup (#8), but it doesn't impact the public API.

**Comment by schrockwell** [comment]
Awesome!
